java-connectors/kafka-connect-gcp-pubsub/src/main/java/io/lenses/streamreactor/connect/gcp/pubsub/source/subscriber/LooselyBoundedQueue.java

```diff
@@ -0,0 +1,167 @@
+ * limitations under the License.
+ */
+package io.lenses.streamreactor.connect.gcp.pubsub.source.subscriber;
+
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Queue;
+
+/**
+ * A LooselyBoundedQueue is a wrapper around a standard {@link Queue} implementation
+ * that introduces a maximum size limit. However, unlike a strictly bounded queue,
+ * it does not fail the {@link #add(Object)} or {@link #addAll(Collection)} operations
+ * when this limit is exceeded. Instead, it logs a debug message when the queue size
+ * exceeds the specified maximum size.
+ *
+ * <p>This class is designed to be used in multi-threaded environments where it is
+ * more important to avoid the failure of add operations than to strictly enforce
+ * a queue size limit. In such scenarios, failing the connector due to queue overflow
+ * is undesirable as it can lead to disruptions in the data processing pipeline.
+ * Hence, this loosely bounded approach allows for some flexibility by permitting
+ * the queue size to exceed the maximum limit while still providing an indication
+ * (through logging) that the limit has been breached.</p>
+ *
+ * @param <X> the type of elements held in this queue
+ */
+@Slf4j
+public class LooselyBoundedQueue<X> implements Queue<X> {
+
+  private final Queue<X> queue;
+  private final int maxSize;
+
+  /**
+   * Constructs a new LooselyBoundedQueue with the specified backing queue and maximum size.
+   *
+   * @param queue   the backing queue
+   * @param maxSize the maximum size of the queue
+   */
+  public LooselyBoundedQueue(Queue<X> queue, int maxSize) {
+    this.queue = queue;
+    this.maxSize = maxSize;
+  }
+
+  @Override
+  public int size() {
+    return queue.size();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return queue.isEmpty();
+  }
+
+  @Override
+  public boolean contains(Object o) {
+    return queue.contains(o);
+  }
+
+  @Override
+  public Iterator<X> iterator() {
+    return queue.iterator();
+  }
+
+  @Override
+  public Object[] toArray() {
+    return queue.toArray();
+  }
+
+  @Override
+  public <T> T[] toArray(T[] a) {
+    return queue.toArray(a);
+  }
+
+  @Override
+  public boolean add(X x) {
+    logIfNoSpareCapacity(1);
+    return queue.add(x);
+  }
+
+  @Override
+  public boolean addAll(Collection<? extends X> c) {
+    logIfNoSpareCapacity(c.size());
+    return queue.addAll(c);
+  }
+
+  @Override
+  public boolean remove(Object o) {
+    return queue.remove(o);
+  }
+
+  @Override
+  public boolean containsAll(Collection<?> c) {
+    return queue.containsAll(c);
+  }
+
+  public boolean hasSpareCapacity(int count) {
+    val newSize = getNewSize(count);
+    return (newSize <= maxSize);
+  }
+
+  private void logIfNoSpareCapacity(int numberOfElements) {
+    final var newSize = getNewSize(numberOfElements);
+    if (newSize > maxSize) {
+      log.debug("Queue will be full - {}/{}", newSize, maxSize);
```

Comment:
The important method is 
```
  public boolean hasSpareCapacity(int count) {
```

as this indicates to the caller if there is space in the queue for additional elements.  This allows us to determine whether to reject additional elements from being presented.

We definitely don't want to cause an error- what are the consequences if there are + 5 additional elements in the queue due to concurrent access?  We definitely don't want to fail the connector for this, we can just reject further messages.

Thank you, I've taken on board your point about extending `ConcurrentLinkedQueue`, instead of implementing the queue interface.  It is unlikely we will need to wrap other queue types.